warehouse/keyword-common/src/main/java/datawave/util/keyword/YakeKeywordExtractor.java

```diff
@@ -0,0 +1,718 @@
+        int pos  = size/2;
+
+        if (size % 2 == 1) return (sentenceOffsets.get(pos));
+        else {
+            int up = sentenceOffsets.get(pos);
+            int down = sentenceOffsets.get(pos-1);
+            return (up+down)/2;
+        }
+    }
+
+    /** Count the digits in a word, a feature used for tagging a word in {@link #calculateTag ( String, int)}.
+     *
+     * @param word the word to analyze.
+     * @param length the maximum number of characters to inspect.
+     * @return the number of digits found in the word as determined by {@link Character#isDigit(char)}.
+     */
+    private static int countDigits(String word, int length) {
+        int count = 0;
+        for (int i = 0; i < length; i++) {
+            if (Character.isDigit(word.charAt(i))) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    /** Count the letters in a word, a feature used for tagging a word in {@link #calculateTag ( String, int)}.
+     *
+     * @param word the word to analyze.
+     * @param length the maximum number of characters to inspect.
+     * @return the number of letters found in the word as determined by {@link Character#isLetter(char)}.
+     */
+    private static int countLetters(String word, int length) {
+        int count = 0;
+        for (int i = 0; i < length; i++) {
+            if (Character.isLetter(word.charAt(i))) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    /** Count the upper case characters in a word, a feature used for tagging a word in {@link #calculateTag ( String, int)}
+     *
+     * @param word the word to analyze.
+     * @param length the maximum number of characters to inspect.
+     * @return the number of upper case letters found in the word as determined by {@link Character#isUpperCase(char)}.
+     */
+    private static int countUpper(String word, int length) {
+        int count = 0;
+        for (int i = 0; i < length; i++) {
+            if (Character.isUpperCase(word.charAt(i))) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    /** Given a list of tokenized sentences, generate a list of candidate keywords from those sentences following the
+     *  specified ngram sizes. Ngrams are generated using a sliding window over the input text. We will always return
+     *  at least one ngram per sentence.
+     *
+     * @param sentences a list of sentences each split into words (should be down-cased).
+     * @param minNGrams the minimum ngram size.
+     * @param maxNGrams the maximum ngram size.
+     * @return a list of ngrams generated from the input sentences ordered based on where they appeared in the input
+     *    sentences. These ngrams are not deduplicated and the words in these ngrams are not normalized in any way.
+     */
+    protected static List<List<String>> calculateNGrams(List<List<String>> sentences, int minNGrams, int maxNGrams) {
```

Comment:
This also generates every ngram between min and max, rather than just the largest ngram possible for each start word.

Reply:
Correct, that is the intent.